src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs

```diff
@@ -0,0 +1,182 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Net.Quic.Implementations.MsQuic.Internal;
+using System.Net.Security;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using static System.Net.Quic.Implementations.MsQuic.Internal.MsQuicNativeMethods;
+
+namespace System.Net.Quic.Implementations.MsQuic
+{
+    internal class MsQuicListener : QuicListenerProvider, IDisposable, IAsyncDisposable
+    {
+        // Functions to invoke in MsQuic
+        private MsQuicApi _api;
+
+        // Security configuration for MsQuic
+        private MsQuicSecurityConfig _secConfig;
+
+        // Pointer to the underlying listener
+        private IntPtr _nativeObjPtr;
+
+        // Handle to this object for native callbacks.
+        private GCHandle _handle;
+
+        // Delegate that wraps the static function that will be called when receiving an event.
+        private ListenerCallbackDelegate _listenerDelegate;
+
+        // Ssl listening options (ALPN, cert, etc)
+        private SslServerAuthenticationOptions _sslOptions;
+
+        // To prevent multiple calls to StartAsync
+        private bool _started;
+
+        private volatile bool _disposed;
+
+        private readonly Channel<MsQuicConnection> _acceptConnectionQueue = Channel.CreateUnbounded<MsQuicConnection>(new UnboundedChannelOptions
```

Comment:
Not sure we'll want to use a channel here (though the usage itself is appropriate). You can reuse the resettable tcs to implement the accepting of connections.

Reply:
How would you be able to enqueue multiple connections at the same time? Ex: you receive 2 connections at the same time, how would you be able to have both be accepted? How would you cap the length?

Reply:
Don't do that, I assume there's a way to enforce back pressure from the MSQuic side (libuv has as well we just didn't implement it that way).

Reply:
I don't think MsQuic does that. @nibanks

Reply:
You can either call stop on the listener when you hit your limit, or just reject new connections.

Reply:
@nibanks The way libuv works is that you get a callback, then you have to call Accept for the callback to be triggered again. We don't want to reject connections at this level, we just want convert this push into pull cleanly with back pressure so we don't buffer too much. Seems like we'll need to call Stop and Start (which I assume is thread safe?) at the right time.

Reply:
If you're listening for new connections, then MsQuic will give you the new connection event when the connection comes in, otherwise it will just fail the connection attempt back to the client. If you are listening, but fail the connection attempt (return an error or don't set the security config in the callback) then it's pretty much the same thing as not listening in the first place (it's a little bit more expensive, but not much).

MsQuic has no buffer for connections that are pending to be indicated up to the listener. If you're listening you get it immediately otherwise the connection is discarded. If you want a buffer for them, you're going to have to implement it at your layer.

Reply:
It sounds to me like we'll need to have some sort of buffer ourselves, and probably have a way to configure the buffer size (i.e. backlog). When the buffer gets full we will need to fail any further connections. I think we could do that just by using a bounded channel here.

Reply:
This is using a bounded channel.